crates/bootstrap_client/src/lib.rs

```diff
@@ -2,24 +2,27 @@
 
 #![deny(missing_docs)]
 
+use base64::Engine;
 use kitsune2_api::{AgentInfoSigned, DynVerifier, K2Error, K2Result, SpaceId};
 use std::sync::Arc;
+use url::Url;
 
 /// Send the agent info, for the given space, to the bootstrap server.
 ///
 /// Note the `blocking_` prefix. This is a hint to the caller that if the function is used in
 /// an async context, it should be treated as a blocking operation.
 pub fn blocking_put(
-    server_url: &str,
+    mut server_url: Url,
     agent_info: &AgentInfoSigned,
 ) -> K2Result<()> {
-    let url = format!(
-        "{server_url}/bootstrap/{}/{}",
-        &agent_info.space, &agent_info.agent
-    );
+    server_url.set_path(&format!(
+        "bootstrap/{}/{}",
+        base64::prelude::BASE64_URL_SAFE_NO_PAD.encode(&**agent_info.space),
```

Comment:
The display implementation can be overridden by the host, which we do in Holochain to print the space id as a DNA hash... Which rather messes this up :)